### PR TITLE
Automocking in test environment

### PR DESCRIPTION
**Pull request by ruibinch**
When running in a test environment, use a mock MMKV instance that utilises an underlying `Map` structure. This allows tests on this package to work without having to manually define mocks or any jest setup files.

Inspired by #367.

### CODE REVIEW

File: src/MMKV.ts [hunk 1]
```diff
@@ -132,9 +134,15 @@ export class MMKV implements MMKVInterface {
    * Creates a new MMKV instance with the given Configuration.
    * If no custom `id` is supplied, `'mmkv.default'` will be used.
    */
-  constructor(configuration: MMKVConfiguration = { id: 'mmkv.default' }) {
+  constructor(
+    configuration: MMKVConfiguration = {
+      id: isJest() ? 'mmkv.mock' : 'mmkv.default',
```
Comment (mrousavy) [inline src/MMKV.ts:139]: This doesn't seem to be required. Can we remove that change?

File: src/PlatformChecker.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export function isJest(): boolean {
+  return Boolean(process.env.JEST_WORKER_ID);
```
Comment (mrousavy) [inline src/PlatformChecker.ts:2]: Wouldn't a `process.env.JEST_WORKER_ID != null` be more semantically correct? Does the linter say anything about this?

File: src/createMMKV.mock.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/* Mock MMKV instance for use in tests */
```
Comment (mrousavy) [inline src/createMMKV.mock.ts:1]: Move that comment down to line 3 so it documents that function

**Review by mrousavy — CHANGES_REQUESTED** [review]
Hey, thanks for your PR, this is awesome!

1. I left a few comments, once we got those resolved we're good otherwise. Code looks clean!
2. Can you confirm that with this change you can now use MMKV without any issues in both actually running the app (no broken behaviour) as well as when running Jest and MMKV instances are being accessible?

Beyond this PR:

1. Do you think we could add a GitHub action (see .github/workflows folder) that runs the Jest tests? Not sure if that really is needed since it doesn't test the actual native MMKV stuff but only the Jest wrapper, but still
2. Do you think it would be possible to test native MMKV functionality as if it was running in an actual RN app with a GitHub action? Maybe spin up a virtual device or something to see if everything's working?

**Comment by mrousavy** [comment]
also CI is failing:

Screenshot 2022-05-05 at 14 10 05

File: src/PlatformChecker.ts (hunk 2 above)
Comment (ruibinch) [reply src/PlatformChecker.ts]: No linter complaints, but agreed that `!= null` captures the meaning better.

**Comment by ruibinch** [comment]
Resolved the comments; let me test this under more situations tomorrow to confirm that everything remains good and update again.

Re other questions:

1. Sounds good, let me add that in a future PR.
2. I've had minimal exposure to GH actions so no idea currently on its capabilities and whether it can support an emulator, but sounds interesting and worth a deeper investigation!

**Comment by ruibinch** [comment]
Tested a couple of scenarios:

1. MMKV continues to work as per normal in an app
2. Running tests, with an open app running an MMKV instance, does not break anything (data remains in the app's MMKV instance)

File: src/MMKV.ts [hunk 4]
```diff
@@ -1,4 +1,6 @@
 import { createMMKV } from './createMMKV';
+import { createMockMMKV } from './createMMKV.mock';
```
Comment (shamilovtim) [inline src/MMKV.ts:2]: I assume you need to use the `require()` API to import this behind an if statement, otherwise you're including it in production builds

File: src/MMKV.ts [hunk 5]
```diff
@@ -1,4 +1,6 @@
...
+import { isJest } from './PlatformChecker';
```
Comment (shamilovtim) [inline src/MMKV.ts:3]: having an export for 1 LOC seems unnecessary. also, file is capitalized despite not being a react component or class.

File: src/MMKV.ts (hunk 4 above)
Comment (ruibinch) [reply src/MMKV.ts:2]: We need to include this in production builds in order for the automocking to work right?

File: src/MMKV.ts (hunk 5 above)
Comment (ruibinch) [reply src/MMKV.ts:3]: Understandable, but I feel this is cleaner than having `process.env.JEST_WORKER_ID != null` inside the `MMKV` constructor.

Re filename - I viewed `PlatformChecker` as more of a "sub-domain" and not a generic file like `helpers`/`utils` hence the capitalisation; not a strong preference though, happy to change it to something else if that's the concensus.

File: src/MMKV.ts (hunk 4 above)
Comment (shamilovtim) [reply src/MMKV.ts:2]: What I'm saying is you're going to include this file in people's JS bundles unless you use the require API and but it behind an if statement

File: src/MMKV.ts (hunk 5 above)
Comment (mrousavy) [reply src/MMKV.ts:3]: I like it the way it is, let's leave it.

File: src/MMKV.ts (hunk 4 above)
Comment (mrousavy) [reply src/MMKV.ts:2]: Is that a fact? We can lazy-load it at the bottom if needed?

Comment (mrousavy) [reply src/MMKV.ts:2]: I mean technically it's not really impacting bundle size, it's just a few lines and no extra dependencies

File: tsconfig.json [hunk 6]
```diff
@@ -1,4 +1,5 @@
 {
+  "exclude": ["example"],
```
Comment (mrousavy) [inline tsconfig.json:2]: Why did you add this here? Now the `example/` doesn't do any typechecking because it doesn't have it's own tsconfig, no?

Comment (ruibinch) [reply tsconfig.json:2]: I added a `example/tsconfig.json` file; `yarn tsc` at the root level throws type errors when it type-checks the `example/**/*.test.ts` files because the Jest types are only defined in `example/node_modules`, so figured it's neater to have separate tsconfig files.

Comment (ruibinch) [reply tsconfig.json:2]: Just realised that the "Run TypeScript in example/" job wasn't running because `typescript` script wasn't defined in `example/package.json` - updated that

**Comment by mrousavy** [comment]
woohoo!

**Comment by mrousavy** [comment]
Released [2.4.0](https://github.com/mrousavy/react-native-mmkv/releases/tag/v2.4.0) 🚀  - thanks for your PR!

**Comment by ruibinch** [comment]
Installing 2.4.0 doesn't include the code changes in `/lib`, though running `yarn build` locally on the current master branch does contain the code changes. Is there some additional config or something else required?

**Comment by ruibinch** [comment]
Cheers, all good now!

**Comment by AshciR** [comment]
This was a joy to read. 

I'm new to mobile development, and I'm learning the ecosystem. I appreciate you enabling developers to write tests for your library.

**Comment by jahirfiquitiva** [comment]
@ruibinch @mrousavy this added an example for a single test... would you mind providing an example or guiding me on how to write multiple tests in multiple files that refer a single mmkv instance shared across all the tests?
